src/gd_avif.c

```diff
@@ -0,0 +1,667 @@
+*/
+
+#define DEFAULT_CHROMA_SUBSAMPLING AVIF_PIXEL_FORMAT_YUV420
+#define DEFAULT_QUANTIZER 30
+#define DEFAULT_QUALITY -1
+#define DEFAULT_SPEED AVIF_SPEED_DEFAULT
+
+// This initial size for the gdIOCtx is standard among GD image conversion functions
+#define NEW_DYNAMIC_CTX_SIZE 2048
+
+// Our quality param ranges from 0 to 100.
+// To calculate quality, we convert from AVIF's quantizer scale, which runs from 63 to 0.
+#define MAX_QUALITY 100
+
+// For computing the number of tiles and threads to use during encoding
+// Maximum threads are from libavif/contrib/gkd-pixbuf/loader.c
+#define MIN_TILE_AREA (512 * 512)
+#define MAX_TILES 6
+#define MAX_THREADS 64
+
+
+/*** Helper function signatures ***/
+
+static avifBool _gdImageAvifCtx (gdImagePtr im, gdIOCtx * outfile, int quality, int speed);
+static void destroyCtxAndAvifIO(avifIO *io);
+static avifIO *createAvifIOFromCtx(gdIOCtx * ctx);
+static avifResult readFromCtx(avifIO * io, uint32_t readFlags, uint64_t offset, size_t size, avifROData * out);
+static void destroyCtxAndAvifIO(avifIO *io);
+static avifBool isAvifError(avifResult result, char * msg);
+static int quality2Quantizer(int quality);
+static uint8_t convertTo8BitAlpha(uint8_t originalAlpha);
+static void setEncoderTilesAndThreads(avifEncoder * encoder, avifRGBImage * rgb);
+
+
+/*****************************************************************************
+ * 
+ *                            DECODING FUNCTIONS
+ * 
+ *****************************************************************************/
+
+/*
+  Function: gdImageCreateFromAvif
+
+    <gdImageCreateFromAvif> is called to load truecolor images from
+    AVIF format files. Invoke <gdImageCreateFromAvif> with an
+    already opened pointer to a file containing the desired
+    image. <gdImageCreateFromAvif> returns a <gdImagePtr> to the new
+    truecolor image, or NULL if unable to load the image (most often
+    because the file is corrupt or does not contain a AVIF
+    image). <gdImageCreateFromAvif> does not close the file.
+
+
+    This function creates a gdIOCtx struct from the file pointer it's passed.
+    And then it relies on <gdImageCreateFromAvifCtx> to do the real decoding work.
+    If the file contains an image sequence, we simply read the first one, discarding the rest.
+
+  Variants:
+
+    <gdImageCreateFromAvifPtr> creates an image from AVIF data
+    already in memory.
+
+    <gdImageCreateFromAvifCtx> reads data from the function
+    pointers in a <gdIOCtx> structure.
+
+  Parameters:
+
+    infile - pointer to the input file
+
+  Returns:
+
+    A pointer to the new truecolor image.  This will need to be
+    destroyed with <gdImageDestroy> once it is no longer needed.
+
+    On error, returns 0.
+*/
+BGD_DECLARE(gdImagePtr) gdImageCreateFromAvif (FILE * infile)
+{
+  gdImagePtr image;
+  gdIOCtx *ctx = gdNewFileCtx(infile);
+
+  if (!ctx) {
+    return GD_FALSE;
+  }
+
+  image = gdImageCreateFromAvifCtx(ctx);
+  // ctx->gd_free(ctx);
```

Comment:
clean up dead comments

although i think you wanted to call this

Reply:
I did want to call this. I commented it out because it gave me an error:

`malloc: *** error for object 0x100d060f0: pointer being freed was not allocated\r\n"`

even though this memory had in fact been allocated by `gdNewFileCtx()`.

Looking around the code, I even see the exact same construction in `gdImageCreateFromJpegEx()`. 

Any idea what I might be missing here?

Reply:
that looks weird.  kind of feels like it might be memory corruption ?

you could try building libavif & gd with sanitizers enabled and run the unittests and see if they catch anything interesting.

Reply:
I haven't done this... but I poked around some more without finding any ideas.

I'd feel bad not releasing this memory, but it's very few bytes....

Reply:
I figured it out. I was already freeing this up in the `destroyCtxAndAvifIO` callback.

That wasn't a good idea. Now that's just `destroyAvifIO`, and we destroy the ctx in the function where it was created.